// artemisa/lib/core/fetch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compileUrl = exports.LOGOUT = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.apiFetch = apiFetch;
exports.fetchOptions = fetchOptions;

var _isomorphicFetch = require('isomorphic-fetch');

var _isomorphicFetch2 = _interopRequireDefault(_isomorphicFetch);

var _service = require('./service');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// HARDCODED: this should be configurable !
// TODO: remove this as its never imported form anywhere?
var LOGOUT = exports.LOGOUT = 'LOGOUT';

var isBrowser = function isBrowser() {
  return (typeof __IS_BROWSER__ === 'undefined' ? 'undefined' : _typeof(__IS_BROWSER__)) !== (typeof undefined === 'undefined' ? 'undefined' : _typeof(undefined));
};
var concatUrl = function concatUrl(base, endUrl) {
  return base + '/' + endUrl;
};
var getBrowserUrl = function getBrowserUrl() {
  return location.protocol + '//' + location.host;
};

function apiFetch(url, options) {
  var baseUrl = !options.baseUrl && isBrowser() ? getBrowserUrl(url) : options.baseUrl || _service.DEFAULT_BASE_URL;
  return (0, _isomorphicFetch2.default)(concatUrl(baseUrl, url), options);
}

var compileUrl = exports.compileUrl = function compileUrl(path) {
  var params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return typeof path === 'string' ? path : path(params);
};

var Method = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT'
};

function fetchOptions() {
  var method = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : Method.GET;
  var params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var authToken = arguments[2];

  var headers = authToken ? { Authorization: 'Bearer ' + authToken } : {};
  var body = undefined;

  if (method === Method.POST || method === Method.PUT) {
    // file here is hardcoded
    if (params.file) {
      headers['Content-Type'] = 'multipart/form-data';
      body = new FormData();
      Object.keys(params).forEach(function (key) {
        if (key !== 'file') {
          body.append(key, params[key]);
        } else {
          body.append(params[key].name, params[key].file);
        }
      });
    } else {
      headers = _extends({}, headers, {
        'Content-Type': 'application/json',
        Accept: 'application/json'
      });
      body = JSON.stringify(params);
    }
  }
  return { method: method, headers: headers, body: body };
}
