// artemisa/lib/core/service.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dataService = exports.DEFAULT_BASE_URL = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _actions = require('./actions');

var _fetch = require('./fetch');

var DEFAULT_BASE_URL = exports.DEFAULT_BASE_URL = 'http://artemisajs.org';

var dataService = exports.dataService = function dataService(configOpts) {
  return function (store) {
    return function (next) {
      return function (action) {
        var r = next(action);
        return (0, _actions.isApiCall)(action) ? callEndpoint(store, action, next, configOpts) : Promise.resolve(r);
      };
    };
  };
};

var callEndpoint = function callEndpoint(store, action, next) {
  var configOpts = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : { call: {} };

  next((0, _actions.onRequestActionCreator)(action));

  return doCallEndpoint(_extends({}, configOpts.call, (0, _actions.callFromAction)(store, action)), next, (0, _actions.onReceiveActionCreator)(action), (0, _actions.onErrorActionCreator)(action), store);
};

// data Service as an object to configure transformations
dataService.transformations = [];
dataService.registerTransformation = function (actionType, fn) {
  dataService.transformations[actionType] = fn;
};

var transformReceiveAction = function transformReceiveAction(receive, store) {
  return _extends({}, receive, {
    data: transform(receive.originType, receive.data, store)
  });
};

var transform = function transform(actionType, value, store) {
  var t = dataService.transformations[actionType];
  return t ? t(value, store) : value;
};

/* eslint-disable no-console */
function doCallEndpoint(callSpec, next, onReceive, onError, store) {
  var method = callSpec.method,
      path = callSpec.path,
      urlParams = callSpec.urlParams,
      body = callSpec.body,
      token = callSpec.token,
      baseUrl = callSpec.baseUrl;

  return (0, _fetch.apiFetch)((0, _fetch.compileUrl)(path, urlParams), _extends({}, (0, _fetch.fetchOptions)(method, body, token), { baseUrl: baseUrl })).then(function (response) {
    return (
      // TODO
      // if (response.status === 401) {
      //   next(unauthorized());
      // }
      response.ok ? response.json().then(function (json) {
        return next(transformReceiveAction(onReceive(json), store));
      })
      // TODO: error parsing should be configurable globally (here we assume that a json object comes
      // with { error: { message } }
      : response.json().then(function (json) {
        return next(onError(json.error.message));
      })
    );
  }).catch(function (error) {
    // console.error('ERROR on API Call', error)
    next(onError(error.message));
  });
}
